cosmo_tester/test_suites/test_manager_upgrade/preupgrade_validations_test.py

```diff
@@ -0,0 +1,312 @@
+                                                properties_dir))
+                fabric.sudo('mv {0} {1}'.format(resources_dir_backup,
+                                                resources_dir))
+
+    def test_node_directories(self):
+        """Validation checks that the required directory structure exists.
+
+        Rename the node_properties and resources directories, check that
+        validation breaks when it notices they're missing, then restore their
+        original names.
+        """
+        blueprint_path = self.get_simple_blueprint()
+        inputs = self.get_upgrade_inputs()
+
+        with self.cfy.maintenance_mode():
+            for node_name in [
+                'influxdb',
+                'rabbitmq',
+                'elasticsearch',
+                'amqpinflux',
+                'logstash',
+                'restservice',
+                'nginx',
+                'riemann',
+                'mgmtworker',
+            ]:
+                with self.move_upgrade_dirs(node_name):
+                    try:
+                        self.cfy.upgrade_manager(
+                            blueprint_path=blueprint_path,
+                            inputs_file=inputs,
+                            validate_only=True)
+                    except sh.ErrorReturnCode as e:
+                        self.assertIn(
+                            'service {0} has no properties file'.format(
+                                node_name),
+                            e.stdout.lower())
+                    else:
+                        self.fail('Upgrade directories validation for {0} '
+                                  'succeeded, but it should have failed - '
+                                  'node properties and rollback directories '
+                                  'are missing'.format(node_name))
+
+    def test_rabbit_credentials_changed(self):
+        """Validation doesn't allow changing rabbitmq credentials.
+
+        Try upgrading with a changed rabbitmq password, check that validation
+        breaks.
+        """
+        rabbitmq_properties = self.get_node_properties('rabbitmq')
+        blueprint_path = self.get_simple_blueprint()
+        with YamlPatcher(blueprint_path) as yamlpatch:
+                yamlpatch.set_value(
+                    ('node_templates.rabbitmq.properties'
+                     '.use_existing_on_upgrade'),
+                    False)
+
+        new_password = rabbitmq_properties['rabbitmq_password'] + '-changed'
+        inputs = self.get_upgrade_inputs(rabbitmq_password=new_password)
+
+        with self.cfy.maintenance_mode():
+            try:
+                self.cfy.upgrade_manager(
+                    blueprint_path=blueprint_path,
+                    inputs_file=inputs,
+                    validate_only=True)
+            except sh.ErrorReturnCode as e:
+                self.assertIn('rabbitmq properties must not change',
+                              e.stdout.lower())
```

Comment:
According to the code, we can validate all 5 immutable properties (rabbitmq_username, rabbitmq_pasword, rabbitmq_endpoint_ip, rabbitmq_ssl_enabled, broker_cert_path) with one iteration. I say why not while we are at it.